SwiftCompilerSources/Sources/Optimizer/Utilities/EscapeInfoWalker.swift

```diff
@@ -0,0 +1,834 @@
+      // the callee analysis.
+      guard let destructors = calleeAnalysis.getDestructors(of: value.type) else {
+        return isEscaping
+      }
+      for destructor in destructors {
+        if destructor.effects.canEscape(argumentIndex: 0, path: p, analyzeAddresses: analyzeAddresses) {
+          return isEscaping
+        }
+      }
+    }
+    return false
+  }
+  
+  /// Handle an apply (full or partial) during the walk-down.
+  private mutating
+  func walkDownCallee(argOp: Operand, apply: ApplySite,
+                      path: Path, state: State) -> Bool {
+    guard let argIdx = apply.argumentIndex(of: argOp) else {
+      // The callee or a type dependent operand of the apply does not let escape anything.
+      return false
+    }
+
+    if canIgnoreForLoadOrArgument(path) { return false }
+
+    // Argument effects do not consider any potential stores to the argument (or it's content).
+    // Therefore, if we need to track stores, the argument effects do not correctly describe what we need.
+    // For example, argument 0 in the following function is marked as not-escaping, although there
+    // is a store to the argument:
+    //
+    //   sil [escapes !%0.**] @callee(@inout X, @owned X) -> () {
+    //   bb0(%0 : $*X, %1 : $X):
+    //     store %1 to %0 : $*X
+    //   }
+    if state.followStores { return isEscaping }
+
+    guard let callees = calleeAnalysis.getCallees(callee: apply.callee) else {
+      // The callees are not know, e.g. if the callee is a closure, class method, etc.
+      return isEscaping
+    }
+
+    let calleeArgIdx = apply.calleeArgIndex(callerArgIndex: argIdx)
+
+    for callee in callees {
+      let effects = callee.effects
+      if !effects.canEscape(argumentIndex: calleeArgIdx, path: path, analyzeAddresses: analyzeAddresses) {
+        continue
+      }
+      if walkDownArgument(calleeArgIdx: calleeArgIdx, argPath: path, state: state,
+                          apply: apply, effects: effects) {
+        return isEscaping
+      }
+    }
+    return false
+  }
+  
+  /// Handle `.escaping` effects for an apply argument during the walk-down.
+  private mutating
+  func walkDownArgument(calleeArgIdx: Int, argPath: Path, state: State,
+                        apply: ApplySite, effects: FunctionEffects) -> Bool {
+    var matched = false
+    for effect in effects.argumentEffects {
+      guard case .escaping(let to, let exclusive) = effect.kind else {
+        continue
+      }
+      if effect.selectedArg.matches(.argument(calleeArgIdx), argPath) {
+        matched = true
+
+        switch to.value {
+        case .returnValue:
+          guard let fas = apply as? FullApplySite, let result = fas.singleDirectResult else { return isEscaping }
+          
+          let state = (exclusive ? state : state.with(knownType: nil)).with(followStores: false)
+          
+          // TODO: doublecheck that only values can be returned
```

Comment:
That's always the case. But it's probably a good idea to add an `assert(result.type.isObject)` here